src/Tokenizer.js

```diff
@@ -489,25 +489,64 @@ module.exports = class Tokenizer {
     }
   }
 
-  strong(src) {
-    const cap = this.rules.inline.strong.exec(src);
+  strong(src, prevChar = '', links) {
+    let cap = this.rules.inline.preStrong.exec(src);
+
     if (cap) {
-      return {
-        type: 'strong',
-        raw: cap[0],
-        text: cap[4] || cap[3] || cap[2] || cap[1]
-      };
+      let text = src;
+      if (links) {
+        links = Object.keys(links);
+        const reg = /(?:\[.*?\]\[.*?\])|(?:\[.*?\](?!\())/g;
+        let match;
+        while ((match = reg.exec(text)) != null) {
+          if (links.includes(match[0].slice(match[0].lastIndexOf('[') + 1, -1))) {
+            text = text.slice(0, match.index) + '[' + 'a'.repeat(match[0].length - 2) + ']' + text.slice(reg.lastIndex);
+          }
+        }
+      }
```

Comment:
Masking every reflink every time we use the `em` or `strong` tokenizers seems like a very brute force way to do it.

Is there a way we could instead do a search for a `*` or `_` and then check if it is in a reflink?

Reply:
Instead of having an `inline.em` rule we could have an `inline.beginEm` and `inline.endEm` rules.

1. check `beginEm` and prevChar
2. search for next `endEm`
3. if current end is part of reflink goto 2

Reply:
> 3. if current end is part of reflink goto 2

The issue with this is that we can't simply check the position of the ending `*`. The pattern must correctly skip or not skip over all reflinks leading up to that point.

Preprocessing the string to handle the reflinks might seem "brute force", but it also nullifies all of them in a single step, rather than iteratively checking each individual reflink for the presence of asterisks or vice-versa, which is a multi-step process that would have to be repeated for each em step anyway.

Reply:
> The issue with this is that we can't simply check the position of the ending *. The pattern must correctly skip or not skip over all reflinks leading up to that point.

That is what the `goto 2` is for. After finding an end that satisfies the other requirements we should check if it is part of a reflink. If it is then we should check for the next end and repeat until we find an end that is not part of a reflink.

It seems to be fine to mask the reflinks for the tests and the benchmarks but if someone has a lot of `em`s and a lot of reflinks it is On^2 and doing a lot of delicate work.

Also, I haven't had time to check, but it is fairly likely that the `inline.em` regex is vulnerable to redos by checking the beginning and end at the same time.

Reply:
I think we are having some miscommunication issues. Is there a possibility of chatting live somewhere to hash out these details so we are on the same page?

Reply:
@calculuschild Sure how would you like to chat? my email is Tony@Brix.ninja if you want to send me something private.